### PR TITLE
PH-37 Fix open REST exporter issues (PLENTY-334)

### PR DESCRIPTION
**Pull request by TheKeymaster**

### CODE REVIEW

File: tests/RegistryServiceTest.php [hunk 1]
```diff
@@ -339,6 +339,79 @@ public function testMissingPluginConfigurationPermissionsAreLoggedAndAllowTheExp
...
+                $pluginSetPluginsResponse
+            )
+        );
+
+        $this->loggerMock->expects($this->once())->method('warning')->with(
+            'Required permission \'Setup > Property > Selection > Show\' has not been granted. ' .
+            'This causes multiSelect properties to not be exported!'
+        );
+
+        $this->registryService->warmUp();
```
Comment (dblaichinger) [inline tests/RegistryServiceTest.php:412]: "...AndExportContinues" 

How does the test ensure that the export continues?

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: src/Response/Entity/Pim/Property/Image.php [hunk 2]
```diff
@@ -27,8 +27,7 @@ class Image extends Entity
     /** @var int */
     private $height;
 
-    /** @var int */
-    private $position;
+    private ?int $position;
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply src/Response/Entity/Pim/Property/Image.php:30]: No this should be an iterative process. That would take way too much time.

File: src/Wrapper/Product.php [hunk 3]
```diff
@@ -215,9 +217,18 @@ protected function processVariations(bool $checkAvailability = true): int
...
 
+            if ($variation->hasCategories()) {
+                $hasCategories = true;
+            }
+
             $variationsProcessed++;
         }
 
+        // If no children have categories, we're skipping this product.
+        if (!$hasCategories) {
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply src/Wrapper/Product.php:228]: A product could have multiple variants. We only want to skip a product if **all variants** have no categories.

File: tests/RegistryServiceTest.php (hunk 1 above)
Comment (TheKeymaster) [reply tests/RegistryServiceTest.php:412]: If it wouldn't continue, an exception would be thrown. Here we explicitly ensure that the line `This causes multiSelect properties to not be exported!` is logged.

File: src/Exporter/Exporter.php [hunk 4]
```diff
@@ -244,4 +246,11 @@ private function getRequiredVariationValues(): array
             'unit'
         ];
     }
+
+    private function getDefaultGuzzleClient(): GuzzleClient
+    {
+        return new GuzzleClient([
+            'verify' => false,
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply src/Exporter/Exporter.php:253]: Yes some Plenty customers (also in the past) and also our x1 demo shop (currently) have had an invalid certificate. Also we do not care if a customer has an invalid certificate or not, we just want to get the data from the REST API.

File: src/Response/Entity/WebStore/Configuration.php [hunk 5]
```diff
@@ -1281,4 +1284,22 @@ public function getUrlTrailingSlash(): ?int
     {
         return $this->urlTrailingSlash;
     }
+
+    /**
+     * @return ItemMeasureUnit[]|string|null
+     */
+    protected function fetchItemMeasureUnit(array $data)
+    {
+        $itemMeasureUnitData = $data['itemMeasureUnit'] ?? null;
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply src/Response/Entity/WebStore/Configuration.php:1293]: `itemMeasureUnit` could not exist at all, therefore using `??` prevents any bugs.

File: src/Wrapper/Product.php [hunk 6]
```diff
@@ -202,7 +204,7 @@ protected function processVariations(bool $checkAvailability = true): int
 
             $position = $variation->getPosition();
             if ($variation->isMain() || !$this->item->getSort()->getValues()) {
-                $this->item->addSort($variation->getPosition());
+                $this->item->addSort($variation->getPosition() ?? 0);
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply src/Wrapper/Product.php:207]: The fun thing is that libflexport default value for sort is `null`, but calling `addSort` requires an `int`. We could not call this method all together, or just set this to `0`. I do not have a strong opinion on either of them.

[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Wrapper/Product.php:207]: I would rather set no sort if we don't have a position. The actual fix should be made in libflexport by setting `?int`.

**Review by dblaichinger — APPROVED** [review]

**Pull request merged**